Add docs on using flatpak as a developer platform (#494)
flatpak/flatpak-docs: docs/flatpak-devel.rst

of a developer workflow.

CI Integration
--------------

Flatpak can be integrated with the CI pipelines and bundles can be
produced as artifacts for quick testing or it can be exported to a
Flatpak repository on a webserver to provide a Nightly Flatpak package.

The CI setup will vary based on the code host and the CI service being
used. A basic workflow for exporting to Gitlab pages is provided in
:doc:`hosting-a-repository` which can be used as a starting point.

`GNOME <https://gitlab.gnome.org/GNOME/Initiatives/-/wikis/DevOps-with-Flatpak#basic-ci>`_
and `KDE <https://develop.kde.org/docs/packaging/flatpak/publishing/#publishing-to-kdes-nightly-repositories>`_
also has their own respective CI setup for doing this, which can be
also be used for inspiration.

`Flatpak Github Actions <https://github.com/flathub-infra/flatpak-github-actions>`_
can be used for GitHub.

Running tests
-------------

Flatpak Builder can also run tests with ``run-tests, test-rule, test-commands, test-args``.

``run-tests: true`` in a module will run tests for that module after
installing.

``test-rule`` is the default target to build when running tests. For

> ``meson`` an ``cmake-ninja`` buildsystems it defaults to ``test``,
> otherwise ``check`` for ``make``. Setting it to an empty string like

Reviewer: ```suggestion
``meson`` and ``cmake-ninja`` buildsystems it defaults to ``test``,
otherwise ``check`` for ``cmake`` and ``autotools``. Setting it to an empty string like
```

Author: thanks

``test-rule: ''`` will disable the target.

``test-commands`` is an array of additional commands that will be run
during tests.

.. code-block:: yaml

  run-tests: true
  test-rule: ''
  test-commands:
     - xvfb-run tests/test_foo

``test-args`` is used to provide finish-args for tests. These do not
affect the normal installation.

.. code-block:: yaml

  build-options:
      test-args:
          - "--socket=x11"
          - "--share=network"

Parallel nigthly and stable applications
----------------------------------------

In general stable and nightly versions or any two parallel branches
of an application should have different application IDs. This is

> mandatory if the application, uses its own well known Dbus name as the
> dbus names need to be distinct.

Reviewer: ```suggestion
mandatory if the application uses its own well-known D-Bus name.
```

Author: changed

> In general stable and nightly versions or any two parallel branches
> of an application should have different application IDs. This is
> mandatory if the application, uses its own well known Dbus name as the
> dbus names need to be distinct.
> 

Reviewer: ```suggestion
In general stable and nightly versions or any two parallel branches
of an application should have different application IDs. This prevents many
potential conflicts such as incompatible configuration files or overlapping well-known D-Bus names.
```

I think the separate config dir is valuable, maybe this phrasing makes sense.

Author: Seems fine.

Author: changed

A standard way is to use the ``.Devel`` suffix to the original Flatpak ID.
The internal application ID, appstream ID, launchable, icon tags,
icon files, desktop files, Dbus service names etc. must follow the new
Flatpak ID for everything to work seamlessly for the user.

This can be integrated with buildsystem `profiles`, so simply building
with ``-Dprofile=Devel`` for example will build the Nightly application
and the default will build the stable appliction.

Buildsystem handling
^^^^^^^^^^^^^^^^^^^^

This is an example for the Meson buildsystem.

The ``src`` subdirectory will contain the main source code of the
application and ``data`` subdirectory will have application metadata
such as desktop file, metainfo file, icons and the ``po`` subdirectory
will have translation files.

The root ``meson.build`` parts should be something like this:

.. code-block::

  i18n = import('i18n')

  profile = get_option('profile')
  if profile == 'development'
    find_program('git')

>     rev_txt = run_command('git','rev-parse','--short','HEAD').stdout().strip()

Reviewer: AFAICT this information is never used?

There is also the `vcs_tag()` function that overlaps with this, depending on needs.

Author: removed.

  else
    rev = ''
    application_id = 'org.example.coolapp'
  endif

  config_h = configuration_data()
  config_h.set_quoted('APPLICATION_ID', application_id)
  config_h.set_quoted('PROFILE', profile)
  config_h.set_quoted('GETTEXT_PACKAGE', 'coolapp')
  config_h.set_quoted('PACKAGE_VERSION', meson.project_version())
  configure_file(
     output: 'config.h',
     configuration: config_h,
  )

  configuration_inc = include_directories('.')

  subdir('data')
  subdir('src')
  subdir('po')

  summary({
    'Profile': get_option('profile'),
  }, section: 'Development')

The root ``meson-options.txt`` should have something like::

  option('profile', type: 'combo', choices: ['default', 'development'], value: 'default')

Now under ``src/meson.build`` or in any subdirectory this can be
specified in a target.

.. code-block::

  coolapp = executable('coolapp', 'main.c',
      include_directories: configuration_inc,
      dependencies: libcoolapp_dep,
      install: true,
  )

Now in ``data/meson.build``, the icon, desktop file and metainfo file
handling can be specified.

Place a desktop template file called ``org.example.coolapp.desktop.in.in``
in ``data/`` with the contents::

> .. code-block::

Reviewer: ```suggestion
.. code-block:: ini
```

Author: fixed

  [Desktop Entry]
  Name=Cool App
  Exec=coolapp
  Icon=@icon@
  Terminal=false
  Type=Application
  Categories=Utility;
  StartupNotify=true

The corresponding meson bits for the desktop file should be

.. code-block::

  desktop_conf = configuration_data()
  desktop_conf.set('icon', application_id)
  desktop_file = i18n.merge_file(
    input: configure_file(
      input: files('org.example.coolapp.desktop.in.in'),
      output: 'org.example.coolapp.desktop.in',
      configuration: desktop_conf,
    ),
    output: '@0@.desktop'.format(application_id),
    type: 'desktop',
    po_dir: '../po',
    install: true,
    install_dir: get_option('datadir') / 'applications',
  )

  desktop_utils = find_program('desktop-file-validate', required: false)
  if desktop_utils.found()
    test('Validate desktop file', desktop_utils, args: [desktop_file])
  endif

Appstream can be handled in the same ``meson.build``. Place a metainfo
template file in ``data/``. Any tag containing part of the application
ID should be templated. Ususally for desktop applications this involves
only the ``id`` and ``launchable`` tag, but the ``icon`` tag, if present
should also use it.

> .. code-block:: yaml

Reviewer: ```suggestion
.. code-block:: xml
```

Author: fixed

  <?xml version="1.0" encoding="UTF-8"?>
  <component type="desktop">
    <id>@appid@</id>
    ...
    <launchable type="desktop-id">@appid@.desktop</launchable>
    ...
  </component>

The corresponding meson.build part for metainfo should be

> .. code-block::

Reviewer: ```suggestion
.. code-block:: meson
```

You could add this to all these code blocks. I'm not sure if our setup has a definition for meson but it is reasonable to assume it could or will some day.

Author: It seems to output a bunch of warnings about lexing literal blocks. I avoided adding it. Not sure if that's because of partial meson.build snippets everywhere.

Reviewer: Bummer.

    install_dir: join_paths(datadir, icondir),
    rename: '@0@.svg'.format(application_id)
  )

  icondir = join_paths('icons', 'hicolor', 'symbolic', 'apps')
  install_data(
    join_paths(icondir, 'org.example.coolapp-symbolic.svg'),
    install_dir: join_paths(datadir, icondir),
    rename: '@0@-symbolic.svg'.format(application_id)
  )

Any gschemea file should also be similarly handled if present.

.. code-block::

  install_data('org.example.coolapp.gschema.xml',
    rename: '@0@.gschema.xml'.format(application_id),
    install_dir: get_option('datadir') / 'glib-2.0' / 'schemas',
  )

Any dbus service file if present should be handled. The ``org.example.coolapp.service.in``
should have ``Name=@appid@``

.. code-block::

  service_conf = configuration_data()
  service_conf.set('appid', application_id)
  service_conf.set('bindir', join_paths(prefix, bindir))
  configure_file(
    input: 'org.example.coolapp.service.in',
    output: '@0@.service'.format(application_id),
    configuration: service_conf,
    install_dir: servicedir
  )

The same pattern can be used for search provider files if present.
The ``org.example.coolapp.service.ini.in`` should look like

> .. code-block::

Reviewer: ```suggestion
.. code-block:: ini
```

Author: fixed

  DesktopId=@appid@.desktop
  BusName=@appid@
  ObjectPath=/org/gnome/AppName@profile@/SearchProvider

and the corresponding meson.build:

.. code-block::

  search_provider_conf = configuration_data()
  search_provider_conf.set('appid', application_id)
  search_provider_conf.set('profile', profile)
  configure_file(
    configuration: search_provider_conf,
    input: files('org.example.coolapp.service.ini.in'),
    install_dir: join_paths(datadir, 'gnome-shell', 'search-providers'),
    output: '@0@.search-provider.ini'.format(application_id)
  )

Application handling
^^^^^^^^^^^^^^^^^^^^

The correct application ID must be supplied when initiating the
application. For example in ``main.c``

.. code-block:: c

>   #include <config.h>

Reviewer: ```suggestion
  #include "config.h"
```

Author: thanks
